### ISSUE TITLE
Create a wrapper to validate API token

### ISSUE THREAD

**Issue by simonaronsson** [containrrr/watchtower#581]
Once #450 is merged, we'd like to add support for wrapping API handles to make them restricted/require a token. For instance, instead of adding a check to each handler, we'd like to be able to do something like this:

```go
authAndHandle := RequireToken(handler.ServeHTTP)
```

See https://github.com/containrrr/watchtower/pull/450#issuecomment-625447646 for any minimal additional details that are available. Also, feel free to reach out to me either here or on gitter to discuss it before you get started on any implementation.

**Comment by m-sedl** [comment]
Hello. I would like to contribute to the creation of the watchtower, but I'm new to the open source community. Could you tell me: did I understand the essence of the task correctly?

Need to create a wrapper for the _handler.ServeHTTP_ object that will execute the following logic?

```go
	authAndHandle := func(w http.ResponseWriter, r *http.Request) {
		// Hijacking the prometheus handler and adding a token check
		if r.Header.Get("Token") != token {
			log.Error("Invalid token. Not serving any metrics.")
			return
		}
		handler.ServeHTTP(w, r)
	}
```

**Comment by victorcmoura** [comment]
Currently, the API package only provides an interface to trigger an update. All of its logic is coded into api.go:
![image](https://user-images.githubusercontent.com/26290053/86264443-e931ce80-bb98-11ea-921b-ba4a1a660941.png)

@simskij, aiming to implement more features into the API, started a new pattern in API package to make it more extensible. In #450, he managed to separate each API feature into one subpackage:
![image](https://user-images.githubusercontent.com/26290053/86264604-1d0cf400-bb99-11ea-8e5e-7a39247417ab.png)

Due to this separation, it would be against proper clean code to keep repeating the following snippet in every subpackage:

```go
if r.Header.Get("Token") != token {
	log.Error("Invalid token. Not serving any metrics.")
	return
}
```

I believe the idea is to do exactly what you proposed. Design a generic token handler to be used in current and future API subpackages to validate the request. But, regarding the best patterns on it, I believe @simskij can make it more clear.

**Comment by m-sedl** [comment]
Thank you very much for your answer. What do you think, in what place of the project is it better to place the future wrapper? For example, in `watchtower/pkg/api/wrappers`. Or will creating a new package be redundant?

**Comment by victorcmoura** [comment]
That's exactly what I thought @simskij could give us guidance. But, if I had to design it myself, I would go for a modification in `registerHandle`, and maybe `registerFunc` as well, by adding a `handler.requireToken` parameter and using some logic within `api.go`.

**Comment by MihailITPlace** [comment]
Good. Then I will wait for @simskij's answer. So far, I can offer such a draft solution to the problem in the `api.go` file. I also remember that it is necessary to delete existing token checks in `metrics.go` and `update.go`.

```go
func (api *API) RequireToken(fn http.HandlerFunc) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		if r.Header.Get("Token") != api.Token {
			log.Error("Invalid token.")
			return
		}
		fn(w, r)
	}
}

// RegisterFunc is a wrapper around http.HandleFunc that also flips the bool used to determine whether to launch the API
func (api *API) RegisterFunc(path string, fn http.HandlerFunc) {
	api.hasHandlers = true
	http.HandleFunc(path, api.RequireToken(fn))
}

// RegisterHandler is a wrapper around http.Handler that also flips the bool used to determine whether to launch the API
func (api *API) RegisterHandler(path string, handler http.Handler) {
	api.hasHandlers = true
	http.Handle(path, api.RequireToken(handler.ServeHTTP))
}
```

**Comment by m-sedl** [comment]
Good. Then I will wait for @simskij's answer. So far, I can offer such a draft solution to the problem in the `api.go` file. I also remember that it is necessary to delete existing token checks in `metrics.go` and `update.go`.

It is worth noting that `metrics_test.go` does not use the logic from `api.go`. Therefore, token verification will not be performed in this test.

I apologize if I made a mistake somewhere.

```go
func (api *API) RequireToken(fn http.HandlerFunc) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		if r.Header.Get("Token") != api.Token {
			log.Error("Invalid token.")
			return
		}
		fn(w, r)
	}
}

// RegisterFunc is a wrapper around http.HandleFunc that also flips the bool used to determine whether to launch the API
func (api *API) RegisterFunc(path string, fn http.HandlerFunc) {
	api.hasHandlers = true
	http.HandleFunc(path, api.RequireToken(fn))
}

// RegisterHandler is a wrapper around http.Handler that also flips the bool used to determine whether to launch the API
func (api *API) RegisterHandler(path string, handler http.Handler) {
	api.hasHandlers = true
	http.Handle(path, api.RequireToken(handler.ServeHTTP))
}
```

**Comment by victorcmoura** [comment]
@MihailITPlace, feel free to fork Watchtower's repo, implement it in the actual source code, and open a pull request. Then someone can review your contribution, make suggestions and then your code can reach the next release =)

**Comment by simonaronsson** [comment]
Excited to have you, @MihailITPlace 🎉  As we currently don't have any public API's, I think the solution you and @victorcmoura have been discussing makes total sense. I'll move over to the PR to comment on specifics (if I can find any).

**Comment by ghost** [comment]
To avoid important communication to get lost in a closed issues no one monitors, I'll go ahead and lock this issue. If you want to continue the discussion, please open a new issue. Thank you! 🙏🏼
